File: plus9qh/submissions/accepted/plus9qh.py
HELLO = "Hello, world!"
LYRICS = [
    "99 bottles of beer on the wall, 99 bottles of beer.",
    "Take one down and pass it around, 98 bottles of beer on the wall.",
    "98 bottles of beer on the wall, 98 bottles of beer.",
    "Take one down and pass it around, 97 bottles of beer on the wall.",
    "97 bottles of beer on the wall, 97 bottles of beer.",
    "Take one down and pass it around, 96 bottles of beer on the wall.",
]


def solve(N: int, X: list[str]):
    # process the text to find a predicted minimal program and predicted quine
    predicted_program = []
    predicted_quine = None
    i = 0
    while i < N:
        if X[i] == HELLO: # is it the output of an H?
            predicted_program.append('H')
            i += 1
        elif X[i] == LYRICS[0]: # is it the output of a 9?
            if X[i:i + 6] == LYRICS:
                predicted_program.append('9')
            else:
                return 'IMPOSSIBLE' # wrong lyrics
            i += 6
        else:
            if all(c in 'HQ9+' for c in X[i]): # if not H or 9, must be a quine
                if predicted_quine == None:
                    predicted_quine = X[i]
                else:
                    if X[i] != predicted_quine:
                        return 'IMPOSSIBLE' # quine contradicts earlier quine
            else:
                return 'IMPOSSIBLE' # invalid quine
            i += 1
    
    # if a quine is found, verify its consistency with the predicted program
    predicted_program = ''.join(predicted_program)
    if predicted_quine != None:
        if predicted_quine.replace('+', '') != predicted_program:
            return 'IMPOSSIBLE' # only difference should be + instructions
        predicted_program = predicted_quine
    
    return predicted_program


def main():
    T = int(input())
    for _ in range(T):
        N = int(input())
        X = [input() for _ in range(N)]
        print(solve(N, X))


if __name__ == '__main__':
    main()

File: plus9qh/submissions/accepted/plus9qh.cpp
#include <iostream>
#include <vector>

using namespace std;

const string HELLO = "Hello, world!";
const string LYRICS[6] = {
    "99 bottles of beer on the wall, 99 bottles of beer.",
    "Take one down and pass it around, 98 bottles of beer on the wall.",
    "98 bottles of beer on the wall, 98 bottles of beer.",
    "Take one down and pass it around, 97 bottles of beer on the wall.",
    "97 bottles of beer on the wall, 97 bottles of beer.",
    "Take one down and pass it around, 96 bottles of beer on the wall."
};

string solve(int N, vector<string> &X) {
    bool okay = true;
    string quine = "";
    string ans = "";
    for (int i = 0; i < N; i++) {
        bool beer = false;
        if (i + 6 <= N) {
            beer = true;
            for (int j = 0; j < 6; j++) {
                if (X[i + j] != LYRICS[j]) {
                    beer = false;
                }
            }
        }
        if (beer) {
            ans.push_back('9');
            i += 5;
        } else if (X[i] == HELLO) {
            ans.push_back('H');
        } else if (!quine.empty() && quine != X[i]) { 
            okay = false;
        } else {
            quine = X[i];
            ans.push_back('Q');
        }
    }
    string clean_program = "";
    for (char c : quine) {
        if (!(c == 'H' || c == 'Q' || c == '9' || c == '+')) {
            okay = false;
        } else if (c != '+') {
            clean_program.push_back(c);
        }
    }
    if (!quine.empty() && clean_program != ans) {
        okay = false;
    }
    if (!quine.empty()) {
        ans = quine;
    }
    return okay ? ans : "IMPOSSIBLE";
}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        int N;
        cin >> N;
        cin.get();
        vector<string> X(N);
        for (int j = 0; j < N; j++) {
            getline(cin, X[j]);
        }
        cout << solve(N, X) << '\n';
    }
}
